### PR TITLE
Image tags

### LINKED ISSUE #7: Image Tags
#### As a user I want to add tags to images.

Acceptance criteria:
- [x] When I add a new image, there is a form field to add tags.
- [x] After I've saved an image with tags, I see those tags displayed with the
  associated image on the index page.
- [x] My application uses the acts-as-taggable-on gem to provide the tagging
  functionality.

Reference material:
- https://github.com/mbleigh/acts-as-taggable-on

Dependencies:
- Image Index

### PR DESCRIPTION
**Pull request by Xin-Tong**
Closes #7

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -20,4 +20,24 @@
     t.datetime "updated_at", null: false
   end
 
+  create_table "taggings", force: :cascade do |t|
```
Comment (shfung) [inline db/schema.rb]: I think more of these fields can have not-null constraints as well.

File: test/controllers/images_controller_test.rb [hunk 2]
```diff
@@ -33,4 +33,19 @@ class ImagesControllerTest < ActionController::TestCase
...
+    params = { image: { name: image_name, url: image_url, tag_list: tag_list } }
+
+    assert_difference 'Image.count', 1 do
+      post :create, params
+    end
+    assert_redirected_to images_path
+    assert_equal image_name, Image.last.name
+    assert_equal image_url, Image.last.url
+    assert_equal tag_list, Image.last.tag_list.join(', ')
+  end
```
Comment (shfung) [inline test/controllers/images_controller_test.rb]: Maybe add another test if the `tag_list` is not a well formed string?

File: test/models/images_test.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+require 'test_helper'
+
+class ImageTest < ActiveSupport::TestCase
+  test 'create with tags' do
+    image = Image.new(url: 'https://google.com', tag_list: 'tag1, tag2')
+    assert_equal 'tag1, tag2', image.tag_list.join(', ')
+  end
```
Comment (shfung) [inline test/models/images_test.rb]: Similar to the previous comment, maybe add another test if the tag_list is not a well formed string?

File: test/flow_tests/images_flow_test.rb [hunk 4]
```diff
@@ -18,4 +18,24 @@ class ImageFlowTest < FlowTestCase
     page.assert_current_path images_path
     assert_selector("img[src=\"#{valid_url}\"]", count: 1)
   end
+
+  test 'create with tags' do
```
Comment (shfung) [inline test/flow_tests/images_flow_test.rb]: Maybe rename the first test to be something like `create without tag`?

File: test/flow_tests/images_flow_test.rb [hunk 5]
```diff
@@ -18,4 +18,24 @@ class ImageFlowTest < FlowTestCase
...
+    valid_url = 'http://doggyfoods.com/wp-content/uploads/2015/12/10.1.jpg'
+    visit root_path
+    click_on('Upload an image')
+    page.assert_current_path new_image_path
+
+    click_on('Main page')
+    page.assert_current_path images_path
```
Comment (shfung) [inline test/flow_tests/images_flow_test.rb]: The above section seems redundant and unnecessary for this test.

File: Gemfile [hunk 6]
```diff
@@ -60,3 +60,6 @@ end
 gem 'browserify-rails'
 
 gem 'simple_form'
+
+#Use ActsAsTaggableOn for addiing tags
```
Comment (shfung) [inline Gemfile]: This comment seems a bit unnecessary.  If you decide to keep it, add a space after the `#` sign, and there is a typo in it.

File: test/controllers/images_controller_test.rb [hunk 7]
```diff
@@ -33,4 +33,19 @@ class ImagesControllerTest < ActionController::TestCase
     assert_equal image_name, Image.last.name
     assert_equal image_url, Image.last.url
   end
+
+  test 'create with tags' do
```
Comment (shfung) [inline test/controllers/images_controller_test.rb]: Maybe rename the `create` test to be something like `create without tag`?

File: test/controllers/images_controller_test.rb [hunk 8]
```diff
@@ -33,4 +33,19 @@ class ImagesControllerTest < ActionController::TestCase
     assert_equal image_name, Image.last.name
     assert_equal image_url, Image.last.url
```
Comment (shfung) [inline test/controllers/images_controller_test.rb]: May add an assertion on the tag_list as well?

File: app/views/images/new.html.erb [hunk 9]
```diff
@@ -1,6 +1,7 @@
 <%= simple_form_for @image do |f| %>
   <%= f.input :name %>
   <%= f.input :url %>
+  <%= f.input :tag_list, label: 'Tags', hint: 'separate by comma'%>
```
Comment (shfung) [inline app/views/images/new.html.erb]: Space before `%>`?

**Comment by Xin-Tong** [comment]
@shfung All the comments have been addressed on https://github.com/AppFolioOnboarding/image-sharer-xin/pull/23/commits/ee66619563906d4ba8fac18b8b5e64620ea069d0 and https://github.com/AppFolioOnboarding/image-sharer-xin/pull/23/commits/44eaba65ba1891575e9e170bed30691fa81106bf. Thanks.

File: Gemfile [hunk 10]
```diff
@@ -60,3 +60,6 @@ end
 gem 'browserify-rails'
 
 gem 'simple_form'
+
+# Use ActsAsTaggableOn for adding tags
```
Comment (rmacklin) [inline Gemfile]: I agree with @shfung, this comment is unnecessary.

File: Gemfile [hunk 11]
```diff
@@ -60,3 +60,6 @@ end
...
+gem 'acts-as-taggable-on', '~> 3.4'
```
Comment (rmacklin) [inline Gemfile]: Minor, but could you add this above `browserify-rails` to keep the gems alphabetized?

File: test/controllers/images_controller_test.rb [hunk 12]
```diff
@@ -30,7 +30,44 @@ class ImagesControllerTest < ActionController::TestCase
...
+    end
+    assert_redirected_to images_path
+
+    image = Image.last
+    assert_equal image_name, image.name
+    assert_equal image_url, image.url
+    assert_equal tag_list, image.tag_list.join(', ')
+  end
+
+  test 'create with tags using empty string' do
```
Comment (rmacklin) [inline test/controllers/images_controller_test.rb]: I'm not sure this test case needs to exist in the controller test

File: test/models/images_test.rb [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+
+  test 'create with tags' do
+    image = Image.new(url: 'https://google.com', tag_list: 'tag1, tag2')
+    assert_equal 'tag1, tag2', image.tag_list.join(', ')
+  end
+
+  test 'create with tags using empty string' do
+    image = Image.new(url: 'https://google.com', tag_list: ' ')
+    assert_empty image.tag_list
+  end
```
Comment (rmacklin) [inline test/models/images_test.rb]: I think you could combine all of these into one `test 'tag_list'` test case, like:
```ruby
image = Image.new(url: 'https://google.com')
assert_empty image.tag_list

image.tag_list = 'tag1, tag2'
assert_equal ['tag1', 'tag2'], image.tag_list

image.tag_list = ' '
assert_empty image.tag_list
```

File: test/controllers/images_controller_test.rb (hunk 12 above)
Comment (Xin-Tong) [inline test/controllers/images_controller_test.rb]: I have covered empty string case in model test. Should I remove this test?

**Comment by Xin-Tong** [comment]
@rmacklin @shfung All the comments have been addressed. I squashed them to keep atomic.

**Comment by shfung** [comment]
Looks good. 👍

**Pull request merged**